IM_calculation/Advanced_IM/advanced_IM_factory.py

```diff
@@ -0,0 +1,116 @@
+)
+COMP_DICT = {"090": 0, "000": 1, "ver": 2}
+
+
+def get_config(config_file=CONFIG_FILE_NAME):
+    with open(config_file) as cf:
+        config = yaml.safe_load(cf)
+    return config
+
+
+def get_im_list(config_file=CONFIG_FILE_NAME):
+    """Retrieves the list of IMs that are present in config_file"""
+    config = get_config(config_file)
+    return list(config.keys())
+
+
+def compute_ims(accelerations, configuration, adv_im_out_dir):
+    """
+    Calculates all the Advanced IMs for the given waveforms
+    :param accelerations: Acceleration array, 1 column for each component. Ordering is specified in COMP_DICT
+    :param configuration: Contains runtime configuration, List of IMs to compute, configuration file containing
+                          location of scripts to run, and path to the open sees binary
+    :return: None (for now)
+    """
+    config = get_config(configuration.config_file)
+    station_name = accelerations.station_name
+
+    with tempfile.TemporaryDirectory() as f:
+        f_dir = os.path.join(f, "")  # ensure has a trailing slash on filename
+        save_waveform_to_tmp_files(f_dir, accelerations, station_name)
+
+        for im in configuration.IM_list:
+            out_dir = os.path.join(adv_im_out_dir, im)
+
+            im_config = config[im]
+            script = [
+                "python",
+                os.path.join(advanced_im_dir, im_config["script_location"]),
+                "--OpenSees_path",
+                configuration.OpenSees_path,
+            ]
+            # waveform component sequence
+            comp_list = ["000", "090", "ver"]
+            script.extend([get_acc_filename(f_dir, station_name, x) for x in comp_list])
+            script.extend([out_dir])
+
+            print(" ".join(script))
+            subprocess.call(script)
+
+
+def get_acc_filename(folder, stat, component):
+    # returns the same filename structure as timeseries.seis2txt
+    return "%s%s.%s" % (folder, stat, component)
+
+
+def save_waveform_to_tmp_files(tmp_folder, accelerations, station_name):
+    """
+    Writes to the 3 files containing values for all components
+    :param acc_file: Dict containing file handles for each component specified
+    :param accelerations: Acceleration array, 1 column for each component. Ordering is specified in COMP_DICT
+    :return: None
+    """
+    # TODO: Fix bug when differing components are specified in calculate_ims
+    for component in COMP_DICT.keys():
+        seis2txt(
+            accelerations.values[:, COMP_DICT[component]],
+            accelerations.DT,
+            tmp_folder,
+            station_name,
+            component,
+        )
+
+def agg_csv(stations, im_calc_dir, im_type):
+    # get csv base on station name
+    # quick check of args format
+    if type(im_type) != str:
+        raise TypeError(
+            "im_type should be a string, but get {} instead".format(type(im_type))
+        )
```

Comment:
Has incorrect type been an issue in the past? Normally I'd just let the path join fail and tell the user to sort it if it's not the right type.
I haven't looked up usages of this function though

Reply:
this can happen if the config wasn't prepared correctly. doesn't really happen often if you are not playing with the config, but there is no drawback to add a check here?

Reply:
I'm not a fan of custom exceptions that aren't caught. It just stops being able to trace the error as well. If you need something like that, an assert may be better suited?